File: com/white/BinaryTreeZigzagLevelOrderTraversal.java
package com.white;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class BinaryTreeZigzagLevelOrderTraversal {

	// 103. Binary Tree Zigzag Level Order Traversal
	
	public static List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        
		ArrayList<List<Integer>> bigList = new ArrayList<>();
		
		if (root==null)
			return bigList;
		
		Stack<TreeNode> stack1 = new Stack<>();
		Stack<TreeNode> stack2 = new Stack<>();
		
		stack1.push(root);
		
		while (!stack1.empty() || !stack2.empty()) {
			
			ArrayList<Integer> smallList = new ArrayList<>();
			
			if (!stack1.empty()) {
				while (!stack1.empty()) {
					TreeNode temp = stack1.pop();
					smallList.add(temp.val);
					if (temp.left!=null)
						stack2.push(temp.left);
					if (temp.right!=null)
						stack2.push(temp.right);
				}
				bigList.add(smallList);
			} else {
				while (!stack2.empty()) {
					TreeNode temp = stack2.pop();
					smallList.add(temp.val);
					if (temp.right!=null)
						stack1.push(temp.right);
					if (temp.left!=null)
						stack1.push(temp.left);
				}
				bigList.add(smallList);
			}
		}
		
		return bigList;
    }
}

File: LeetcodeCore/BinaryTreeZigzagLevelOrderTraversal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetcodeCore
{
    public class BinaryTreeZigzagLevelOrderTraversal
    {
        // 103. Binary Tree Zigzag Level Order Traversal
        public IList<IList<int>> ZigzagLevelOrder(TreeNode root)
        {
            var resultList = new List<IList<int>>();
            if (root == null)
                return resultList;
            var stack1 = new Stack<TreeNode>();
            var stack2 = new Stack<TreeNode>();
            var flag = true;    // flag can be extracted as an input, to determine which way of zigzag it goes
            stack1.Push(root);

            while (stack1.Count != 0 || stack2.Count != 0)
            {
                if (stack1.Count != 0)
                {
                    resultList.Add(TraverseALevel(stack1, stack2, flag).ToList());
                    flag = !flag;
                }
                else if (stack2.Count != 0)
                {
                    resultList.Add(TraverseALevel(stack2, stack1, flag).ToList());
                    flag = !flag;
                }
            }

            return resultList;
        }

        private IEnumerable<int> TraverseALevel(Stack<TreeNode> giveStack, Stack<TreeNode> takeStack, bool flag)
        {
            while (giveStack.Count != 0)
            {
                var node = giveStack.Pop();
                if (flag)
                {
                    if (node.left != null) takeStack.Push(node.left);
                    if (node.right != null) takeStack.Push(node.right);
                }
                else
                {
                    if (node.right != null) takeStack.Push(node.right);
                    if (node.left != null) takeStack.Push(node.left);
                }
                yield return node.val;
            }
        }
    }
}
